test/integration/test_dataplane_scale_subresource.go

```diff
@@ -105,4 +85,80 @@ func TestDataPlaneScaleSubresource(t *testing.T) {
 
 	t.Log("verifying dataplane scales to 3 replicas")
 	require.Eventually(t, testutils.DataPlaneHasNReadyPods(t, GetCtx(), dataplaneName, clients, 3), testutils.GatewayReadyTimeLimit, testutils.ObjectUpdateTick)
+
+	t.Log("simulating kubectl rollout restart by adding restart annotation")
+	deploymentCopy := deployment.DeepCopy()
+	if deploymentCopy.Spec.Template.Annotations == nil {
+		deploymentCopy.Spec.Template.Annotations = make(map[string]string)
+	}
+
+	restartTime := time.Now()
+	deploymentCopy.Spec.Template.Annotations["kubectl.kubernetes.io/restartedAt"] = restartTime.Format(time.RFC3339)
+
+	err = clients.MgrClient.Patch(GetCtx(), deploymentCopy, client.MergeFrom(&deployment))
+	require.NoError(t, err)
+
+	t.Log("waiting for new ReplicaSet to be created")
+
+	var newReplicaSet *appsv1.ReplicaSet
+	require.Eventually(t, func() bool {
+		rs := FindReplicaSetNewerThan(
+			t,
+			GetCtx(),
+			clients.MgrClient,
+			restartTime,
+			namespace.Name,
+			dataplane,
+		)
+		if rs != nil {
+			t.Logf("Found new ReplicaSet %s created at %v", rs.Name, rs.CreationTimestamp)
+			newReplicaSet = rs
+			return true
+		}
+		return false
+	}, testutils.GatewayReadyTimeLimit, testutils.ObjectUpdateTick, "Failed to find new ReplicaSet after rollout restart")
+
+	t.Log("verifying new ReplicaSet has correct replica count")
+	assert.NotNil(t, newReplicaSet, "New ReplicaSet should be created")
+
+	// Wait for the ReplicaSet to have the correct replica count
+	require.Eventually(t, func() bool {
+		if err := clients.MgrClient.Get(GetCtx(), types.NamespacedName{
+			Namespace: newReplicaSet.Namespace,
+			Name:      newReplicaSet.Name,
+		}, newReplicaSet); err != nil {
+			t.Logf("Error getting ReplicaSet: %v", err)
+			return false
+		}
+		t.Logf("Current ReplicaSet replicas: %d", *newReplicaSet.Spec.Replicas)
+		return *newReplicaSet.Spec.Replicas == 3
+	}, testutils.GatewayReadyTimeLimit, testutils.ObjectUpdateTick, "New ReplicaSet should have 3 replicas")
+
+	t.Log("waiting for rollout to complete")
+	require.Eventually(t, func() bool {
+		if err := clients.MgrClient.Get(GetCtx(), types.NamespacedName{
+			Namespace: newReplicaSet.Namespace,
+			Name:      newReplicaSet.Name,
+		}, newReplicaSet); err != nil {
+			return false
+		}
+		return newReplicaSet.Status.ReadyReplicas == *newReplicaSet.Spec.Replicas
+	}, testutils.GatewayReadyTimeLimit, testutils.ObjectUpdateTick, "New ReplicaSet should have all replicas ready")
+
+	t.Log("verifying dataplane still has 3 ready replicas after rollout")
+	require.Eventually(t, testutils.DataPlaneHasNReadyPods(t, GetCtx(), dataplaneName, clients, 3), testutils.GatewayReadyTimeLimit, testutils.ObjectUpdateTick)
+
+	// Test the specific issue where scaling after restart didn't work properly
+	t.Log("scaling DataPlane to 4 replicas after restart")
+	// Test scaling through the scale subresource
+	t.Log("testing scaling through scale subresource")
```

Comment:
This is not technically scaling through the subresource but rather through the `.replicas` spec field, right? If so I'd update the comment here.

Reply:
Yes. I should conduct this test here using `kubectl scale` or through the `scale` API.

Reply:
👍 Let's do this then.